pkg/referrerstore/oras/oraswithcache.go

```diff
@@ -0,0 +1,133 @@
+package oras
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/cespare/xxhash/v2"
+	"github.com/deislabs/ratify/pkg/common"
+	"github.com/deislabs/ratify/pkg/ocispecs"
+	"github.com/deislabs/ratify/pkg/referrerstore"
+	"github.com/deislabs/ratify/pkg/referrerstore/config"
+	"github.com/dgraph-io/ristretto"
+	"github.com/dgraph-io/ristretto/z"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+)
+
+type Operation int
+
+const (
+	Name Operation = iota
+	GetConfig
+	ListReferrers
+	GetBlobContent
+	GetReferenceManifest
+	GetSubjectDescriptor
+)
+
+type orasStoreWithInMemoryCache struct {
+	base      referrerstore.ReferrerStore
+	cache     *ristretto.Cache
+	cacheConf *OrasCacheConf
+}
+
+type OrasCacheConf struct {
+	Enabled bool `json:"cacheEnabled"`
+	Ttl     int  `json:"ttl"`
+}
+
+type orasStoreFactoryWithCache struct{}
+
+func keyToHash(key interface{}) (uint64, uint64) {
+	if key == nil {
+		return 0, 0
+	}
+	switch k := key.(type) {
+	case string:
+		return z.MemHashString(k), xxhash.Sum64String(k)
+	default:
+		return 0, 0
+	}
+}
+
+// Create creates a new oras store decorated with in-memory cache to cache
+// results of ListReferrers API.
+func (s *orasStoreFactoryWithCache) Create(storeBase referrerstore.ReferrerStore, cacheConf *OrasCacheConf) (referrerstore.ReferrerStore, error) {
+	memoryCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 1e5,
+		MaxCost:     1 << 30,
+		BufferItems: 64,
+		KeyToHash:   keyToHash,
+	})
```

Comment:
Is `orasStoreWithInMemoryCache` a singleton? or other?

I've concern that there is a memory leak on invoking `ristretto.NewCache()` as `memoryCache` is never garbage collected by calling `memoryCache.Close()`.

Reply:
That's a good question. `orasStoreWithInMemoryCache` and base `orasStore` are not singleton. They're created during service initialization or CRD updated. Furthermore, the Oras store can be created multiple times if different configs are provided. I feel the case we need to take care is that when we delete store CRD, it should also clean up the store resources.

Reply:
As discussed offline, there might be 4 ways to avoid memory leak:
1. Add Close() API to referrerStore so that we can override the Close API to call cache's Close function. But it needs to update the Store API Spec.
2. Use SetFinalizer to let garbage collector clean it up. But there is no guarantee that the finalizer will be called or when.
3. Use some cache library without explicit Close() function. I've listed some libraries in the top description.
4. Refactor to use singleton Cache. In this case, new store might have the old cache and cannot update the TTL.

Reply:
IMO, I feel option 3 and 4 are better than 1 and 2 as we could have more control on GC and no need to modify API sepc. For option 3 we could use [freecache](https://github.com/coocood/freecache), and for option 4, we can stick with ristretto. And in terms of the control on the total capacity and entry size, I would vote for ristretto.

Reply:
@toddysm is it a valid use case that multiple oras stores are set up? And are we OK to make those stores to share the same cache? And in the multi-tenant case, I think we would need to separate caches for each tenant.

If we do need to consider the above cases, I feel I have to adopt the option 3.

Reply:
The final decision is to stick with ristretto.